[Experimental] Product Filters Overlay Navigation block

dinhtungdu opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR adds a new block called `Navigation`, which is an inner block of `Product Filters Overlay`. The block contains a button to close the overlay. This PR only touches the block creation, editor settings, and editor preview, the frontend side of the block will be worked on in future PRs.

The Navigation block added in this PR:
- Can only be added inside `Product Filters Overlay` block.
- Can change the justification of inner text and icon.
- Can be sticky to the top of the overlay while scrolling.
- Styling:
  - By default, the block inherits the current theme style.
  - Typography, Color, Dimensions, and Border settings are available.
  - There are three navigation styles: Label and icon, Label only, and Icon only.
  - There are three button styles: Link, Fill, and Outline.
  - Icon size can be changed using a custom control.

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #47975 

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This should be tested by developers only.
0. Ensure the experimental blocks are enabled:
    - If you use the WooCommerce Beta Tester plugin, you can enable the experimental blocks in Tools > WCA Test Helper > Features. Toggle the `experimental-blocks` option.
    - If you build this PR locally, build the branch with `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build` command to have experimental blocks available.
1. Edit the Filters Overlay template part.
2. See the Navigation (Experimental) block.
3. Try changing settings and see the editor preview reflects those changes:
   - Alignment
   - Position
   - Color/Typo/Border/Dimensions.
   - Style, Button Style, Icon Size.
   - For Icon Size, verify the unit selector works as expected.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

dinhtungdu committed:
add: block settings and editor preview

dinhtungdu committed:
add: block scaffold

dinhtungdu committed:
chore: discard change to pnpm lock file

dinhtungdu committed:
add navigation block to the template part

dinhtungdu committed:
test: add basic E2E tests

dinhtungdu committed:
chore: changelog

dinhtungdu committed:
fix: no full width button

dinhtungdu committed:
test: remove layout check

dinhtungdu committed:
tweak: dynamically fetch button styles

roykho commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters-overlay/block.json:
```diff
@@ -32,6 +32,19 @@
 		"overlayPosition": {
 			"type": "string",
 			"default": "left"
+		},
```
Does adding the style here not affect the frontend? The gap issue identify from before is in the editor.

dinhtungdu replied:
@roykho Yes, it applies to the frontend, but I don't see a reason why we shouldn't match the editor style with the frontend style.

roykho replied:
So the fix originally was just for the editor as commented from here https://github.com/woocommerce/woocommerce/pull/48472#pullrequestreview-2116811799

I don't know if this fix is needed in the front end or not as we're not currently working on the frontend.

roykho reviewed (commented):
Mostly looking good. I just want to point out the icon size controls are not smooth. I think it should function like the fontsizepicker where the element resizes in real-time on-fly. You can just delay the setAttributes. But if we decide to switch to using the fontsizepicker, then this comment is moot.

dinhtungdu committed:
try: fontsizepicker

dinhtungdu committed:
remove: FontSizePicker dute to lack of ability to change label

dinhtungdu committed:
fix: use correct icon from the design

dinhtungdu commented:
@roykho thanks for the review. I updated the icon size control to make it smoother when changing the size using the slider. I also gave FontSizePicker a try but discarded it because we can't set a custom label for the control. The label is fixed to `Size`.

> You can just delay the setAttributes. But if we decide to switch to using the fontsizepicker, then this comment is moot.

We have to do so even with FontSizePicker, you can try the default Size under the Typography panel, it's laggy when you change the size using the slider.

Can you please give this one another look?

roykho commented:
> We have to do so even with FontSizePicker, you can try the default Size under the Typography panel, it's laggy when you change the size using the slider.

Hmmm, when I tried it, it wasn't laggy. See video:


https://github.com/woocommerce/woocommerce/assets/2132595/d737edc7-18ca-42b9-b2b8-b73cbac56823

However if you say we cannot change the label, then that's another story.

roykho commented:
If we decide to use this custom icon size component, can you please move it up so that it can be reused? We will need it also in the work I am doing [here](https://github.com/woocommerce/woocommerce/pull/48732/)

dinhtungdu commented:
> Hmmm, when I tried it, it wasn't laggy. See video:
@roykho it's interesting that it's smooth in the Post Editor but not Site Editor: https://cloudup.com/cMvRznXhVL6

> If we decide to use this custom icon size component, can you please move it up so that it can be reused?

Actually, now I'm not sure that we need a unit switcher for icon size, px is probably enough. @jarekmorawski should we allow the Icon Size to have different units other than `px`, like rem or em?

thealexandrelara commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters-overlay-navigation/components/size-control.tsx:
```diff
@@ -0,0 +1,73 @@
+		value ? value.replace( /[^a-zA-Z]/g, '' ) : 'px'
+	);
+
+	useEffect( () => {
+		onChange( numericSize, unit );
+	}, [ numericSize, unit, onChange ] );
```
I believe it would be better to avoid using ‘useEffect’ to trigger the ‘onChange’ function here. Doing so can cause performance issues because every time the component re-renders, the ‘onChange’ function is created again. When React performs a shallow comparison of the array that is passed, it will not be the same, and this will re-trigger multiple times. Sometimes, this can even cause infinite loops. 

```suggestion
```

You can call the `onChange` function on each place that you are changing the "numericSize" and/or the "unit" (I'll leave some examples in the code below).

thealexandrelara commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters-overlay-navigation/components/size-control.tsx:
```diff
@@ -0,0 +1,73 @@
+				<UnitControl
+					value={ `${ numericSize }${ unit }` }
+					units={ units }
+					onChange={ ( newSize: string ) => {
+						setNumericSize( parseInt( newSize, 10 ) );
+						if ( ! newSize ) {
+							setUnit( 'px' );
+							return;
+						}
+						if ( ! newSize.includes( unit ) ) {
+							setUnit( newSize.replace( /[^a-zA-Z]/g, '' ) );
+						}
+					} }
```
For example:

```suggestion
					onChange={ ( newSize: string ) => {
					    const newNumericSize = parseInt( newSize, 10 );
					    let newUnit = 'px';
						setNumericSize( newNumericSize );
						if ( ! newSize ) {
							setUnit( newUnit );
							onChange( newNumericSize, newUnit );
							return;
						}
						if ( ! newSize.includes( unit ) ) {
						    newUnit = newSize.replace( /[^a-zA-Z]/g, '' );
							setUnit( newUnit );
						}
						onChange( newNumericSize, newUnit );
					} }
```

thealexandrelara commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters-overlay-navigation/components/size-control.tsx:
```diff
@@ -0,0 +1,73 @@
+				/>
+				<RangeControl
+					value={ numericSize }
+					onChange={ ( newSize: number ) =>
+						setNumericSize( newSize )
+					}
```
And here:

```suggestion
					onChange={ ( newSize: number ) =>
						setNumericSize( newSize )
						onChange( newSize, unit );
					}
```

thealexandrelara reviewed (changes requested):
Thank you for working on this, Tung! I tested it, and it is working as expected. I left some suggestions in the code because they might be related to the performance issues Roy reported. 

I noticed that the icon size doesn't seem to be affected by the Typography settings, so I'd like to confirm if this is expected:


https://github.com/woocommerce/woocommerce/assets/20469356/763a3220-8224-4e2a-9f15-f54df3038651

---

One thing I realized from the instructions is that is says "... you can enable the experimental blocks in Settings > WCA Test Helper > Features. Toggle the experimental-blocks option.". I believe the correct path is: "Tools > WCA Test Helper > Features", right?

dinhtungdu committed:
address review feedback

dinhtungdu commented:
@thealexandrelara @roykho in [b2fcb77](https://github.com/woocommerce/woocommerce/pull/48677/commits/b2fcb7783ac59030659e7def58c64408399c867a):

- I updated the icon to use the font size if the icon size isn't set.
- I changed to use RangeControl as Roy does in another PR, let's use what is available in the core first until we see demand.
- I used `setAttributes` to control the icon size as Core does with [the FontSizePicker component](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/font-size.js#L100-L113).
- I also updated the testing instructions as Alexandre suggested.


##### About the performance issue

As shared in [a previous comment](https://github.com/woocommerce/woocommerce/pull/48677#issuecomment-2189265858), there is a difference between Site Editor and Post Editor in terms of smoothness when changing the font size using the FontSizePicker slider. In 9057351d706754a91432dd2ec4432f2ae83f1bd8, I had to use a state to control the font size and debounce the attribute update to make the size change smooth. But I decided to revert those changes to use `setAttributes` as Core does.

I will have to dig into the difference between Site and Post Editor more and report that issue if it's valid and hasn't been reported yet.

roykho commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters-overlay-navigation/edit.tsx:
```diff
@@ -0,0 +1,172 @@
+	return (
+		<nav
+			className={ clsx(
+				'wc-block-product-filters-overlay-navigation-wrapper',
```
I think we should follow the standard convention of naming classes like `wc-block-product-filters-overlay-navigation__wrapper`

roykho commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters-overlay-navigation/edit.tsx:
```diff
@@ -0,0 +1,172 @@
+						/>
+					) }
+
+					<RangeControl
```
We should probably hide this control if style is set to `Label only`.

roykho commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters-overlay-navigation/edit.tsx:
```diff
@@ -0,0 +1,172 @@
+					) }
+
+					<RangeControl
+						className="wc-block-product-filters-overlay-navigation-icon-size-control"
```
Same about the naming convention mentioned above. Should probably be `wc-block-product-filters-overlay-navigation__icon-size-control`

dinhtungdu committed:
update class to follow naming convention. use explicit value to better indicate the navigation style

dinhtungdu commented:
@roykho I updated the class naming and hid the icon size control when the style is set to label only following your suggestions. Please give this one another look, thanks!

roykho commented:
Sorry, I found another issue. The global styles colors (text,background) don't seem to be applying.

![Google Chrome - Filters Overlay ‹ Template part ‹ woo-blocks-dev ‹ Editor — WordPress 2024-06-26 at 7 23 54 AM](https://github.com/woocommerce/woocommerce/assets/2132595/eca72bb0-ebc9-4ef2-80ca-5d0a3c43cd42)

dinhtungdu committed:
fix: remove unnecessary style that breaks the global style

dinhtungdu commented:
@roykho thanks for the finding, I fixed the global style in 8e1c7fe

roykho commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters-overlay-navigation/style.scss:
```diff
@@ -0,0 +1,16 @@
+}
+
+.wc-block-product-filters-overlay-navigation {
+	display: inline-flex !important;
```
Since this also effects the frontend, we should try to avoid using `!important` as much as possible. Does this need to be `inline-flex`? Will `flex` not work here? If you really need it to be `inline-flex`, you can just add specificity like `.is-layout-flex.wc-block-product-filters-overlay-navigation { display: inline-flex; }`. This way it will be easier for themes to override this.

dinhtungdu replied:
Great catch! I check it again and we don't need inline-flex. It's left from my previous implementation. I removed it in 655aa15, thanks a lot!

roykho reviewed (commented):
I just left a small comment about a css issue. Also one thing I thought about is if the site is RTL (right-to-left), perhaps the icon should switch sides. `X Close` instead of `Close X`.

dinhtungdu committed:
fix: remove unnecessary style

dinhtungdu commented:
@roykho thanks for the heads up on RTL, I didn't think about it at all. I tested the block with the test store running an RTL language, and the icon is automatically changed to the left side of the text. I guess it may related to default flex direction. It means that the block is RTL support already. Can you give this one another look?

<details><summary>Screenshot</summary>
<p>

![image](https://github.com/woocommerce/woocommerce/assets/5423135/46e3b47e-d960-4696-9eaa-dd6f92124d4b)


</p>
</details>

roykho reviewed (approved):
Thanks for the changes. It's looking good to me!

thealexandrelara reviewed (approved):
Thank you for addressing all the raised points. PR approved 🚀
